python/modules/TopReweightProducer.py

```diff
@@ -0,0 +1,148 @@
+
+        ## Description of variables
+        ## Stop0l_topptWeight is the full reweighting correction with the top pT + Pow/MG
+        ## Stop0l_topMGPowWeight is only the Pow/MG correction
+        ## Stop0l_topptOnly is only the top Pt reweighting that is done on gentops
+        ## Stop0l_topptOnly_Up/Down is the systematic variation using inputs from Zhenbin
+        ## Mathematically the relation is Stop0l_topptWeight = Stop0l_topptOnly * Stop0l_topMGPowWeight
+        if not self.isData:
+            self.out.branch('Stop0l_topptWeight', 			"F")
+            self.out.branch('Stop0l_topMGPowWeight', 			"F")
+            self.out.branch('Stop0l_topptOnly', 		        "F")
+            self.out.branch('Stop0l_topptOnly_Up', 			"F")
+            self.out.branch('Stop0l_topptOnly_Down', 			"F")
+
+    def topPTSyst(self, toppt):
+        ptrange = [0, 25, 75, 125, 200, 250, 300, 350, 450, 600]
+        pt_index = -1
+        y_up = []
+        y_dn = []
+        x = []
+        syst_out = []
+        for i in xrange(len(ptrange)):
+            if toppt < float(ptrange[i]):
+                if toppt >= 25: pt_index = i - 1
+                else:           pt_index = 1
+                x = [ptrange[pt_index + 1], ptrange[pt_index]]
+                y_up = [self.topPTSyst_up.GetBinContent(self.topPTSyst_up.GetXaxis().FindBin(ptrange[pt_index + 1])), self.topPTSyst_up.GetBinContent(self.topPTSyst_up.GetXaxis().FindBin(ptrange[pt_index]))]
+                y_dn = [self.topPTSyst_dn.GetBinContent(self.topPTSyst_dn.GetXaxis().FindBin(ptrange[pt_index + 1])), self.topPTSyst_dn.GetBinContent(self.topPTSyst_dn.GetXaxis().FindBin(ptrange[pt_index]))]
+                break
+            else:
+                pt_index = len(ptrange) - 1
+                x = [ptrange[pt_index], ptrange[pt_index - 1]]
+                y_up = [self.topPTSyst_up.GetBinContent(self.topPTSyst_up.GetXaxis().FindBin(ptrange[pt_index])), self.topPTSyst_up.GetBinContent(self.topPTSyst_up.GetXaxis().FindBin(ptrange[pt_index - 1]))]
+                y_dn = [self.topPTSyst_dn.GetBinContent(self.topPTSyst_dn.GetXaxis().FindBin(ptrange[pt_index])), self.topPTSyst_dn.GetBinContent(self.topPTSyst_dn.GetXaxis().FindBin(ptrange[pt_index - 1]))]
+
+        def line(pt, m, c):
+            return m*np.clip(pt, 0, 800) + c
+
+        coef_up = np.polyfit(x, y_up, 1)
+        coef_dn = np.polyfit(x, y_dn, 1)
+        up = line(toppt, coef_up[0], coef_up[1])
+        dn = line(toppt, coef_dn[0], coef_dn[1])
+
+        if up < 0.5 or dn < 0.5: print("toppt: {0}, coef_up: {1}, coef_dn: {2}, x: {3}, y_up: {4}, y_dn: {5}, ptrange: {6}".format(toppt, coef_up, coef_dn, x, y_up, y_dn, ptrange[pt_index]))
+
+        return up, dn
+
+    def topPTWeight(self, genparts):
+        genTops = []
+        genTops_up = []
+        genTops_dn = []
+        mgpowheg = []
+        for gp in genparts:
+            if gp.statusFlags & 8192 == 0: continue
```

Comment:
Remind me again what is this 8192?

Reply:
Its to check if the genpart is "isLastCopy" so that the momentum used is correct. And yes I have tested this many times, but more tests couldn't hurt.